core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java

```diff
@@ -39,41 +42,77 @@ O initContextAndExecuteWithFallback(
+            if (initFuture.isDone()) {
+                // Initialization has been done immediately.
+                final boolean success;
+                try {
+                    success = initFuture.get();
+                } catch (Exception e) {
+                    throw Exceptions.peel(e);
                 }
 
-                // No need to call `fail()` because failed by `DefaultRequestContext.init()` already.
-                final Throwable cause = ctx.log().partial().requestCause();
-                assert cause != null;
-                return fallbackResponseFactory.apply(ctx, cause);
+                return initContextAndExecuteWithFallback(delegate, ctx, errorResponseFactory, success);
+            } else {
+                return futureConverter.apply(initFuture.handle((success, cause) -> {
+                    try {
+                        if (cause != null) {
+                            throw cause;
+                        }
+
+                        return initContextAndExecuteWithFallback(
+                                delegate, ctx, errorResponseFactory, success);
+                    } catch (Throwable t) {
+                        fail(ctx, t);
+                        return errorResponseFactory.apply(ctx, t);
+                    }
+                }));
             }
         } catch (Throwable cause) {
             fail(ctx, cause);
-            return fallbackResponseFactory.apply(ctx, cause);
+            return errorResponseFactory.apply(ctx, cause);
+        }
+    }
+
+    private static <I extends Request, O extends Response, U extends Client<I, O>>
+    O initContextAndExecuteWithFallback(
+            U delegate, DefaultClientRequestContext ctx,
+            BiFunction<ClientRequestContext, Throwable, O> errorResponseFactory, boolean succeeded)
+            throws Exception {
+
+        if (succeeded) {
+            return pushAndExecute(delegate, ctx);
+        } else {
+            // Context initialization has failed, which means:
+            // - ctx.log() has been completed with an exception.
+            // - ctx.request() has been aborted (if not null).
+            // - the decorator chain was not invoked at all.
+            // See `init()` and `failEarly()` in `DefaultClientRequestContext`.
+
+            // Call the decorator chain anyway so that the request is seen by the decorators.
+            final O res = pushAndExecute(delegate, ctx);
+
+            // We will use the fallback response which is created from the exception
+            // raised in ctx.init(), so the response returned can be aborted.
+            if (res instanceof StreamMessage) {
+                ((StreamMessage<?>) res).abort();
+            }
+
+            // No need to call `fail()` because failed by `DefaultRequestContext.init()` already.
+            final Throwable cause = ctx.log().partial().requestCause();
+            assert cause != null;
```

Comment:
Not related to this PR. Could we propagrate the `cause` to `res` via `res.abort(cause)`?

Reply:
Done :-)